make DType a dataclass

thelonejordan opened the pull request:
- [x] Make `DType` a dataclass

* directly using `DType` object as keys creates issues while subclassing the dataclass, so used the `name` attribute as key instead.
* added support for `PtrDType` to `DType` comparison. (this was not needed for `NamedTuple`).

chenyuxyz commented:
making DType a dataclass is good, but i don't see the point of making np a property. the goal is to remove .np!

thelonejordan commented:
> making DType a dataclass is good, but i don't see the point of making np a property. the goal is to remove .np!

@chenyuxyz the goal was to remove numpy incrementally. Removing numpy is a big change in itself https://github.com/tinygrad/tinygrad/pull/3075. I was trying to limit this PR to just converting NamedTuple to dataclass. We can always delete later.

chenyuxyz commented:
is making .np a property a requirement to make DType a dataclass?

thelonejordan commented:
> is making .np a property a requirement to make DType a dataclass?

@chenyuxyz It's not. There was a TODO to remove np from attributes. So I removed it and added "np" as a property to support the .np calls. The functionality is exactly the same, the numpy to dtype mapping is handled dynamically, so cleaner.

thelonejordan commented on tinygrad/dtype.py:
```diff
@@ -70,9 +73,9 @@ def fields() -> Dict[str, DType]: return DTYPES_DICT
 
   # NOTE: these are image dtypes
   @staticmethod
-  def imageh(shp): return ImageDType(100, 2, "imageh", np.float16, shp, dtypes.float32)
```
this change is ambiguous!

thelonejordan replied:
fixed!

thelonejordan commented:
> is making .np a property a requirement to make DType a dataclass?

@chenyuxyz reverted this.

geohot commented:
Cool!
